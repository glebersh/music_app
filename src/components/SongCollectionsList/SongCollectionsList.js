import React from 'react';
import { useSelector } from 'react-redux';
import SongCollectionCard from '../SongCollectionCard/SongCollectionCard';

const SongCollectionsList = ({ type }) => {

  const content = useSelector(state => {
    if (type === 'albums') {
      return state.searchReducer.searchResult.albums.items;
    }
    else if (type === 'playlists') {
      return state.searchReducer.searchResult.playlists.items;
    }
  });


  const songCollectionData = content.map((item) =>
  (<SongCollectionCard key={item.data.uri} name={item.data.name} type={type} owner={type === 'playlists' ? item.data.owner.name : null}
    coverURL={type === 'albums' ? item.data.coverArt.sources[0].url : item.data.images.items[0].sources[0].url}
    id={item.data.uri} artist={type === 'albums' ? item.data.artists.items[0].profile.name : null}
    artistID={type === 'albums' ? item.data.artists.items[0].uri.replace('spotify:artist:', '') : null} />)
  )

  return (
    <>
      {songCollectionData}
    </>
  )
};

export default SongCollectionsList;