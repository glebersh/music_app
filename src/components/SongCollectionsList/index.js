import SongsCollectionList from './SongsCollectionList';
export default SongsCollectionList;