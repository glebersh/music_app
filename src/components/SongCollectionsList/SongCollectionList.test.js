import SongCollectionsList from "./SongCollectionsList";
import { render } from "@testing-library/react";
import '@testing-library/jest-dom';
import userEvent from "@testing-library/user-event";

import * as reduxHooks from 'react-redux';
import { BrowserRouter } from "react-router-dom";

jest.mock('react-redux');
const mockedUseSelector = jest.spyOn(reduxHooks, 'useSelector');
const mockedUseDispatch = jest.spyOn(reduxHooks, 'useDispatch');

export const mockedAlbumData = [{
  artists: [
    {
      id: '1z4g3DjTBBZKhvAroFlhOM',
      name: 'Red Velvet',
    }
  ],
  id: '58OR7UoaJkJzqeQGClHzh1',
  images: [
    {
      height: 300,
      url: 'https://i.scdn.co/image/ab67616d00001e02d2ef237da7f94762997c2083',
      width: 300
    },
  ],
  name: '‘The ReVe Festival 2022 - Birthday’',
  total_tracks: 5,
  type: 'album',
  uri: 'spotify:album:58OR7UoaJkJzqeQGClHzh1'
}];

export const mockedPlaylistData = [{
  collaborative: false,
  description: 'SEULGI SOLO',
  external_urls: {
    spotify: 'https://open.spotify.com/playlist/111Tmphrp0a1xc4RmTvLkw'
  },
  href: 'https://api.spotify.com/v1/playlists/111Tmphrp0a1xc4RmTvLkw',
  id: '111Tmphrp0a1xc4RmTvLkw',
  images: [
    {
      height: null,
      url: 'https://i.scdn.co/image/ab67706c0000bebb3fdbcea2f13fd7327a178f7c',
      width: null
    }
  ],
  name: 'Red Velvet All Songs - In order',
  owner: {
    display_name: 'alejulula',
    external_urls: {
      spotify: 'https://open.spotify.com/user/ivf0a92xhjsa4upuwiewnbgs7'
    },
    href: 'https://api.spotify.com/v1/users/ivf0a92xhjsa4upuwiewnbgs7',
    id: 'ivf0a92xhjsa4upuwiewnbgs7',
    type: 'user',
    uri: 'spotify:user:ivf0a92xhjsa4upuwiewnbgs7'
  },
  primary_color: null,
  'public': null,
  snapshot_id: 'NzcsMTk5ZDJmNGY4NmRjNjFjN2NlYThlMWMyNmU0MDNhMGU1ZGJiMWRlYw==',
  tracks: {
    href: 'https://api.spotify.com/v1/playlists/111Tmphrp0a1xc4RmTvLkw/tracks',
    total: 149
  },
  type: 'playlist',
  uri: 'spotify:playlist:111Tmphrp0a1xc4RmTvLkw'
}];


describe('Song list tests', () => {
  it('Should render and match snapshot if loading is \'resovled\'', () => {
    mockedUseSelector
      .mockReturnValueOnce(mockedAlbumData)
      .mockReturnValueOnce('resolved')
      .mockReturnValueOnce(null);

    const songCollection = render(
      <BrowserRouter>
        <SongCollectionsList collectionType='ALBUMS' />
      </BrowserRouter>);
    expect(songCollection).toMatchSnapshot();
  });

  it('If loading is \'loading\' should render spinner', () => {
    mockedUseSelector
      .mockReturnValueOnce(mockedAlbumData)
      .mockReturnValueOnce('loading')
      .mockReturnValueOnce(null);

    const songCollection = render(
      <BrowserRouter>
        <SongCollectionsList collectionType='ALBUMS' />
      </BrowserRouter>);
    expect(songCollection.getByRole('spinner')).toBeInTheDocument();
  });

  it('Should render alert if error', () => {
    mockedUseSelector
      .mockReturnValueOnce(mockedAlbumData)
      .mockReturnValueOnce('resolved')
      .mockReturnValueOnce(true);

    const songCollection = render(
      <BrowserRouter>
        <SongCollectionsList collectionType='ALBUMS' />
      </BrowserRouter>);
    expect(songCollection.getByRole('error-alert')).toBeInTheDocument();
  });

  it('Should render if playlist', () => {
    mockedUseSelector
      .mockReturnValueOnce(mockedPlaylistData)
      .mockReturnValueOnce('resolved')
      .mockReturnValueOnce(false);

    const songCollection = render(
      <BrowserRouter>
        <SongCollectionsList collectionType='PLAYLISTS' />
      </BrowserRouter>);
    expect(songCollection.getByTestId('albums-list')).not.toBeEmptyDOMElement();
  });
});