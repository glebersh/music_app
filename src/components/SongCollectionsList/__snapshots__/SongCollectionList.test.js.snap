// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Song list tests Should render and match snapshot if loading is 'resovled' 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        data-testid="albums-list"
      >
        <div
          class="css-1ics6b3"
          data-testid="album-container"
        >
          <div
            class="css-79elbk"
          >
            <img
              class="album-cover css-1psm24t"
              loading="lazy"
              src="https://i.scdn.co/image/ab67616d00001e02d2ef237da7f94762997c2083"
            />
            <i
              class="bi bi-play-fill song-collection-icon"
            />
          </div>
          <p
            class="chakra-text css-16ngg4e"
          >
            ‘The ReVe Festival 2022 - Birthday’
          </p>
          <a
            href="/artists/1z4g3DjTBBZKhvAroFlhOM"
          >
            <p
              class="chakra-text css-1hlu8je"
            >
              Red Velvet
            </p>
          </a>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      data-testid="albums-list"
    >
      <div
        class="css-1ics6b3"
        data-testid="album-container"
      >
        <div
          class="css-79elbk"
        >
          <img
            class="album-cover css-1psm24t"
            loading="lazy"
            src="https://i.scdn.co/image/ab67616d00001e02d2ef237da7f94762997c2083"
          />
          <i
            class="bi bi-play-fill song-collection-icon"
          />
        </div>
        <p
          class="chakra-text css-16ngg4e"
        >
          ‘The ReVe Festival 2022 - Birthday’
        </p>
        <a
          href="/artists/1z4g3DjTBBZKhvAroFlhOM"
        >
          <p
            class="chakra-text css-1hlu8je"
          >
            Red Velvet
          </p>
        </a>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
